src/docs/05-go-client/19-workflow-non-deterministic-error.md

```diff
@@ -0,0 +1,178 @@
+
+![change-workflow-ownership](../../shared/img/non-determinsitic-errors-page-visuals/change-workflow-ownership.png)
+
+Workflow A then will be picked up by Host B and continues its execution. This process is called <b>change of workflow ownership</b>. However, after Host B gains ownership of the Workflow A, it does not have any information about its historical executions. For example, Workflow A may have executed many activities and it fails. Host B needs to redo all its history until the moment of failure. The process of reconstructing history of a workflow is called <b>history replay</b>.
+
+In general, any errors occurs during the replay process are called <b>non-deterministic errors</b>. We will explore different types of non-deterministic errors in sections below but first let's try to understand how Cadence is able to perform the replay of workflow in case of failure.
+
+## Decision tasks of workflow
+
+In the previous section, we learned that Cadence is able to replay workflow histories in case of failure. We will learn exactly how Cadence keeps track of histories and how they get replayed when necessary.
+
+Workflow histories are built based on event-sourcing, and each history event are persisted in Cadence storage. In Cadence, we call these history events <b>decision tasks</b>, the foundation of history replay. Most decision tasks have three status - <b>Scheduled</b>, <b>Started</b>, <b>Completed</b> and we will go over decision tasks produced by each Cadence operation in section below.
+
+When changing a workflow ownership of host and replaying a workflow, the decision tasks are downloaded from database and persisted in memory. Then during the workflow replaying process, if Cadence finds a decision task already exists for a particular step, it will immediately return the value of a decision task instead of rerunning the whole workflow logic. Let's take a look at the following simple workflow implementation and explictly list all decision tasks produced by this workflow.
+
+```go
+func SimpleWorkflow(ctx workflow.Context) error {
+	ao := workflow.ActivityOptions{
+		...
+	}
+	ctx = workflow.WithActivityOptions(ctx, ao)
+
+	var a int
+	err := workflow.ExecuteActivity(ctx, ActivityA).Get(ctx, &a)
+	if err != nil {
+		return err
+	}
+
+	workflow.Sleep(time.Minute)
+
+	err = workflow.ExecuteActivity(ctx, ActivityB, a).Get(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+    workflow.Sleep(time.Hour)
+	return nil
+}
+```
+
+In this workflow, when it starts, it first execute ActivityA and then assign the result to an integer. It sleeps for one minute and then use the integer as an input argument to execute ActivityB. Finally it sleeps for one hour and completes.
+
+The following table lists the decision tasks stack produced by this workflow. It may looks overwhelming first but if you associate each decision task with its corresponding Cadence operation, it becomes self-explanatory.
```

Comment:
```suggestion
The following table lists the decision tasks stack produced by this workflow. It may look overwhelming first but if you associate each decision task with its corresponding Cadence operation, it becomes self-explanatory.
```
tbh I'd probably recommend tossing the contents of all this into a google doc or something, and browsing through the spelling and grammar suggestions.